File: tests/go/tests-go-testing-allure3-playwright/schema.go
package tests

import (
	"embed"
	"encoding/json"
	"fmt"
	"math"
	"testing"

	"github.com/stretchr/testify/require"
)

//go:embed schemas/*.json
var schemaFS embed.FS

func AssertSchema(t *testing.T, raw []byte, name string) {
	t.Helper()
	schemaRaw, err := schemaFS.ReadFile("schemas/" + name)
	require.NoError(t, err, name)
	var schema map[string]any
	require.NoError(t, json.Unmarshal(schemaRaw, &schema))
	var data any
	require.NoError(t, json.Unmarshal(raw, &data), string(raw))
	require.NoError(t, validateJSON(data, schema), name)
}

func validateJSON(data any, schema map[string]any) error {
	if typ, ok := schema["type"].(string); ok {
		if err := checkType(data, typ); err != nil {
			return err
		}
	}
	switch typed := data.(type) {
	case map[string]any:
		if req, ok := schema["required"].([]any); ok {
			for _, key := range req {
				name, _ := key.(string)
				if _, ok := typed[name]; !ok {
					return fmt.Errorf("missing required %q", name)
				}
			}
		}
		props, _ := schema["properties"].(map[string]any)
		if additional, ok := schema["additionalProperties"].(bool); ok && !additional {
			for key := range typed {
				if _, ok := props[key]; !ok {
					return fmt.Errorf("unexpected property %q", key)
				}
			}
		}
		for key, prop := range props {
			value, ok := typed[key]
			if !ok {
				continue
			}
			propSchema, _ := prop.(map[string]any)
			if err := validateJSON(value, propSchema); err != nil {
				return fmt.Errorf("%s: %w", key, err)
			}
		}
	case []any:
		itemSchema, _ := schema["items"].(map[string]any)
		if itemSchema == nil {
			return nil
		}
		for i, item := range typed {
			if err := validateJSON(item, itemSchema); err != nil {
				return fmt.Errorf("[%d]: %w", i, err)
			}
		}
	case string:
		if min, ok := schema["minLength"].(float64); ok && len(typed) < int(min) {
			return fmt.Errorf("string shorter than minLength %v", min)
		}
	}
	return nil
}

func checkType(data any, typ string) error {
	switch typ {
	case "object":
		if _, ok := data.(map[string]any); !ok {
			return fmt.Errorf("want object, got %T", data)
		}
	case "array":
		if _, ok := data.([]any); !ok {
			return fmt.Errorf("want array, got %T", data)
		}
	case "string":
		if _, ok := data.(string); !ok {
			return fmt.Errorf("want string, got %T", data)
		}
	case "integer":
		n, ok := data.(float64)
		if !ok || n != math.Trunc(n) {
			return fmt.Errorf("want integer, got %T", data)
		}
	case "number":
		if _, ok := data.(float64); !ok {
			return fmt.Errorf("want number, got %T", data)
		}
	case "boolean":
		if _, ok := data.(bool); !ok {
			return fmt.Errorf("want boolean, got %T", data)
		}
	}
	return nil
}

func Message(t *testing.T, body map[string]any) string {
	t.Helper()
	msg, _ := body["message"].(string)
	require.NotEmpty(t, msg)
	return msg
}

File: tests/rust/tests-rust-testing-reqwest-selenium/src/schema.rs
use serde_json::Value;

pub fn assert_schema(raw: &[u8], name: &str) {
    let schema_raw = match name {
        "auth-response.json" => include_str!("../schemas/auth-response.json"),
        "error.json" => include_str!("../schemas/error.json"),
        "health.json" => include_str!("../schemas/health.json"),
        "items.json" => include_str!("../schemas/items.json"),
        "profile.json" => include_str!("../schemas/profile.json"),
        other => panic!("unknown schema {other}"),
    };
    let schema: Value = serde_json::from_str(schema_raw).expect(name);
    let data: Value = serde_json::from_slice(raw).unwrap_or_else(|_| {
        panic!("body is not JSON for {name}: {}", String::from_utf8_lossy(raw))
    });
    validate_json(&data, &schema).unwrap_or_else(|err| panic!("{name}: {err}"));
}

fn validate_json(data: &Value, schema: &Value) -> Result<(), String> {
    if let Some(typ) = schema.get("type").and_then(Value::as_str) {
        check_type(data, typ)?;
    }
    match data {
        Value::Object(map) => {
            if let Some(required) = schema.get("required").and_then(Value::as_array) {
                for key in required {
                    let name = key.as_str().unwrap_or_default();
                    if !map.contains_key(name) {
                        return Err(format!("missing required {name:?}"));
                    }
                }
            }
            let props = schema
                .get("properties")
                .and_then(Value::as_object)
                .cloned()
                .unwrap_or_default();
            if schema
                .get("additionalProperties")
                .and_then(Value::as_bool)
                == Some(false)
            {
                for key in map.keys() {
                    if !props.contains_key(key) {
                        return Err(format!("unexpected property {key:?}"));
                    }
                }
            }
            for (key, prop) in &props {
                if let Some(value) = map.get(key) {
                    validate_json(value, prop).map_err(|err| format!("{key}: {err}"))?;
                }
            }
        }
        Value::Array(items) => {
            if let Some(item_schema) = schema.get("items") {
                for (i, item) in items.iter().enumerate() {
                    validate_json(item, item_schema).map_err(|err| format!("[{i}]: {err}"))?;
                }
            }
        }
        Value::String(text) => {
            if let Some(min) = schema.get("minLength").and_then(Value::as_u64) {
                if (text.len() as u64) < min {
                    return Err(format!("string shorter than minLength {min}"));
                }
            }
        }
        _ => {}
    }
    Ok(())
}

fn check_type(data: &Value, typ: &str) -> Result<(), String> {
    let ok = match typ {
        "object" => data.is_object(),
        "array" => data.is_array(),
        "string" => data.is_string(),
        "integer" => data.as_i64().is_some() || data.as_u64().is_some(),
        "number" => data.is_number(),
        "boolean" => data.is_boolean(),
        _ => true,
    };
    if ok {
        Ok(())
    } else {
        Err(format!("want {typ}, got {data}"))
    }
}

pub fn message(body: &serde_json::Map<String, Value>) -> String {
    let msg = body
        .get("message")
        .and_then(serde_json::Value::as_str)
        .unwrap_or("");
    assert!(!msg.is_empty(), "message");
    msg.to_string()
}
